termwiz/src/render/terminfo.rs

```diff
@@ -52,32 +52,37 @@ impl TerminfoRenderer {
         allow(clippy::cyclomatic_complexity, clippy::cognitive_complexity)
     )]
     fn flush_pending_attr<W: UnixTty + Write>(&mut self, out: &mut W) -> Fallible<()> {
-        macro_rules! attr_on_off {
-            ($cap_on:ident, $cap_off:ident, $attr:expr, $accesor:ident, $sgr:ident) => {
-                let value = $attr.$accesor();
-                if value != self.current_attr.$accesor() {
-                    let out = out.by_ref();
-                    let on: bool = value.into();
-                    if on {
-                        if let Some(attr) = self.get_capability::<cap::$cap_on>() {
-                            attr.expand().to(out)?;
-                        } else {
-                            write!(out, "{}", CSI::Sgr(Sgr::$sgr(value)))?;
-                        }
-                    } else {
-                        if let Some(attr) = self.get_capability::<cap::$cap_off>() {
-                            attr.expand().to(out)?;
-                        } else {
-                            write!(out, "{}", CSI::Sgr(Sgr::$sgr(value)))?;
-                        }
-                    }
+        macro_rules! attr_on {
+            ($cap:ident, $sgr:expr) => {
+                if let Some(attr) = self.get_capability::<cap::$cap>() {
+                    attr.expand().to(out.by_ref())?;
+                } else {
+                    write!(out, "{}", CSI::Sgr($sgr))?;
                 }
             };
+            ($sgr:expr) => {
+                write!(out, "{}", CSI::Sgr($sgr))?;
+            };
         }
 
         if let Some(attr) = self.pending_attr.take() {
+            let mut current_foreground = self.current_attr.foreground;
+            let mut current_background = self.current_attr.background;
+
             if !attr.attribute_bits_equal(&self.current_attr) {
-                if let Some(sgr) = self.get_capability::<cap::SetAttributes>() {
+                // Updating the attribute bits also resets the colors.
+                current_foreground = ColorAttribute::Default;
+                current_background = ColorAttribute::Default;
+
+                // The SetAttributes capability can only handle single underline and slow blink.
```

Comment:
This looks like it has some unintentionally broad impact; the intent of the prior code was to use the terminfo-provided SetAttributes sequences for as much as possible, falling back on our manual mostly ECMA-48 CSI sequences.

With this change, if we have single-underline or slow-blink then we won't use the terminfo at all.
While in most cases this will be a wash, I wouldn't be surprised if there was some terminal out there that wasn't totally ECMA-48 and this would be a tricky issue to run down.

I think that we should probably have this logic:

* If cap::SetAttributes is available, use it to set the attributes that it supports, then manually do the remaining attributes "by hand"
* Otherwise do it "by hand"

Reply:
If we have single-underline or slow-blink we should still use the terminfo-based SetAttributes.  It's just the existence of double-underline or rapid-blink that makes us fall back to the old behaviour.